bee-ternary/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+//! A general-purpose ternary manipulation, translation and encoding crate.
+//!
+//! # Features
+//!
+//! - Creation of trit and tryte buffers with multiple encodings
+//! - Safe encoding API that allows the efficient manipulation and sharing of trit and tryte buffers and slices
+//! - Mutation of trit buffers and slices
+//! - Ternary BigInt implementation
+//! - Balanced and unbalanced ternary
+//! - `serde` support
+//!
+//! # Encodings
+//!
+//! This crate includes support for many different trit encodings. Encodings allow the trading off
+//! of different features against each other.
+//!
+//! [`T1B1`] is the canonical default encoding and represents every trit with a single byte of
+//! memory. It is the fastest encoding to manipulate since no bitwise operations are necessary to
+//! pack and unpack it from memory during manipulation. As a result of this, it also permits
+//! certain extra features like mutable chunking and accessing its contents through ordinary
+//! slices.
+//!
+//! [`T3B1`] is also commonly used. It provides good compression and has the advantage that it has
+//! an identical bit representation as a [`Tryte`] slice. For this reason, it is the only encoding
+//! that can be converted to a tryte slice with no overhead.
+//!
+//! [`T5B1`] is the most compressed encoding. It provides very high storage densities (almost
+//! optimal, in fact) and is the densest encoding supported by this crate.
+//!
+//! It is likely that one of the 3 encodings above will suit your requirements. In addition, this
+//! crate also supports [`T2B1`] and [`T4B1`] for completeness.
```

Comment:
I would not add things to the crate that have no usecase. Quite frankly, the reason "for completeness" is not a good-enough reason for me. Btw, this is a tough statement. There are other encodings possible, and even used in PoW, which I call the T8B2 encoding, where 1 trit is not encoded by 2 consecutive bits in memory, but in two separate bytes at the same index. It has the same density as a T4B1, but a different memory layout so that you can parallelize bit operations.

Reply:
And thinking about it, there is also the `T9B2`, which has really interesting properties (it is the only encoding where tryte and byte lengths match, and was used in the Ict prototype)

Reply:
The traits in `raw` are publicly exposed such that these representations can be implemented if necessary.

Regarding the `T2B1` and `T4B1` encodings, I'd argue that *many* things in this crate don't have a clearly specified use-case: it's a utility crate so it's designed to be applicable to a variety of scenarios.

Reply:
well, it's almost certain that T2B1 and T4B1 won't ever be used, because they have no single property that outshines all the others.
T1B1: most easy to manipulate
T3B1: most readable
T5B1: most dense
T8B2: most parallizable
T9B2: most performant *and* dense

Reply:
@thibault-martinez What's your view on this?

Reply:
I think he's fine with how it is. I don't think those 2 encodings are well enough justified other than they were easily possible. We will have them as baggage IMHO, and need to maintain them, but no one seems to really care. As I said for each of those T2B1 and T4B1 there is a better alternative, hence they are superseded from the get go. Having said that I am fine with keeping them if no one else than me cares :smile:

Reply:
we can close this then. The rest of the PR looks very good to me!

Reply:
I do not have a strong stance on this matter. While I acknowledge the fact that the probability of these is very low, if not null, I don't have a specific issue with these existing. You decide.